neqo-transport/src/pmtud.rs

```diff
@@ -0,0 +1,180 @@
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use std::{cell::RefCell, net::IpAddr, rc::Rc};
+
+use neqo_common::{qdebug, qtrace};
+
+use crate::{recovery::SentPacket, Stats};
+
+// From https://datatracker.ietf.org/doc/html/rfc1191#section-7.1, with a few modifications.
+const MTU_SIZES: [usize; 13] = [
```

Comment:
Overall, I think that the steps here are a bit questionable.  There are some tiny steps and some huge ones.   Is there some empirical research that might inform the steps we use?  RFC 1191 and the RFCs you cite are all ancient.

Reply:
@gorryfair had a paper a few years back: https://ieeexplore.ieee.org/document/8506538. I wonder if we should just do a binary search (to within 4-16 bytes) on the range [1280..1500].

Reply:
<img width="1026" alt="Screenshot 2024-05-22 at 09 34 07" src="https://github.com/mozilla/neqo/assets/200328/cfeaa1d5-1933-410b-80c8-2f7c4091af01">
Note this is TCP MSS, so add mentally headers.

Reply:
That's useful.  I'd start by uncompressing the 1280 barrier for v6, which leads to a very similar graph overall on both.

It suggests that there isn't a whole lot of value in probing below 1380 or so (our current value).  Given how few people are in that bucket, we might ignore anything below 1350 or so, like we have done, though if we are able to do continuous checking as I suggested, having 1200 on the list as a backstop would deal with those odd few cases where we start at 1350 and the path changes.

Beyond that, we should probe at just a few steps.  I see one at ~1420 (I'm correcting for TCP header, adjust accordingly) for v4 that looks like a good point to check, then ~1460 for v6, and finally 1500 for both.  Beyond that, I don't see much value in doing anything other than a few points (16k and 64k, and maybe add 4k).  In practice, we're not going to see that often, outside of loopback tests.  Besides, the faster we can hit the 64k number, the better.

Given these numbers, I'm not seeing much value in a binary search.  The graph is just too flat to bother.  That region of slope is narrow enough that trying to dial in an exact number there doesn't seem to be worth it.